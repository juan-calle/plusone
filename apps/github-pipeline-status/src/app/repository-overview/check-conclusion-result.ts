import { CheckConclusionState } from '@plusone/github-schema'

export const CheckConclusionResult: Record<CheckConclusionState, string> = {
  ACTION_REQUIRED: '\u26a0',
  CANCELLED: '',
  FAILURE: '\u2717',
  NEUTRAL: '',
  SKIPPED: '',
  STALE: '',
  STARTUP_FAILURE: '',
  TIMED_OUT: '',
  SUCCESS: '\u2713',
}
