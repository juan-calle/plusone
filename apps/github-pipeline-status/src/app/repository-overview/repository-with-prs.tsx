import React from 'react'
import {
  CheckConclusionState,
  Commit,
  Repository,
} from '@plusone/github-schema'

type RepositoryWithPRsProps = Repository

const CheckConclusionResult: Record<CheckConclusionState, string> = {
  ACTION_REQUIRED: '\u26a0',
  CANCELLED: '',
  FAILURE: '\u2717',
  NEUTRAL: '',
  SKIPPED: '',
  STALE: '',
  STARTUP_FAILURE: '',
  TIMED_OUT: '',
  SUCCESS: '\u2713',
}

export const RepositoryWithPrs: React.FC<RepositoryWithPRsProps> = ({
  name,
  url,
  defaultBranchRef,
  pullRequests: { totalCount: prCount, nodes },
}) => {
  const prCheckState = nodes
    .flatMap((node) => node.commits.nodes)
    .flatMap((node) => node.commit.checkSuites.nodes)
    .flatMap((node) => node.conclusion)

  const prsWithSuccessfulCheck = prCheckState.filter(
    (state) => state === CheckConclusionState.Success,
  ).length
  const prsWithFailedCheck = prCheckState.filter(
    (state) => state === CheckConclusionState.Failure,
  ).length
  const prsWithOtherCheckState = prCheckState.filter(
    (state) =>
      state !== CheckConclusionState.Success &&
      state !== CheckConclusionState.Failure,
  ).length

  const defaultBranchCheckConclusion = (defaultBranchRef.target as Commit).checkSuites.nodes
    .flatMap((node) => node.conclusion)
    .pop()

  return (
    <tr>
      <td>
        <a href={url}>{name}</a>
      </td>
      <td>
        {`${defaultBranchRef.name} ${
          CheckConclusionResult[defaultBranchCheckConclusion] ?? ''
        }`}
      </td>
      <td>{prCount > 0 ? <a href={url + '/pulls'}>({prCount} PRs)</a> : ''}</td>
      <td>{prsWithSuccessfulCheck > 0 && prsWithSuccessfulCheck}</td>
      <td>{prsWithFailedCheck > 0 && prsWithFailedCheck}</td>
      <td>{prsWithOtherCheckState > 0 && prsWithOtherCheckState}</td>
    </tr>
  )
}
