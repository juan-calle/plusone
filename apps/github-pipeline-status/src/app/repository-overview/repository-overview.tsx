import React, { useState } from 'react'
import {
  Organization,
  PageInfo,
  Repository,
  SearchType,
} from '@plusone/github-schema'
import { useGitHubPagination } from '@plusone/hooks'
import { useQuery } from 'react-query'

import { useOctokit } from '../octokit-provider/octokit-provider'

import { RepositoryWithPrs } from './repository-with-prs'

const PAGE_SIZE = 20

export interface RepositoryOverviewProps {
  organizationName: Organization['name']
}

export const RepositoryOverview = ({
  organizationName,
}: RepositoryOverviewProps) => {
  const octokit = useOctokit()

  const {
    pages,
    onSuccess,
    nextPage,
    prevPage,
    getPageRequest,
  } = useGitHubPagination(PAGE_SIZE)

  const [queryString, setQueryString] = useState<string>('')

  const { data, isLoading } = useQuery(
    ['repositories', organizationName, queryString, pages.currentPage],
    async () => {
      return await octokit.graphql<{
        search: {
          repositoryCount: number
          pageInfo: PageInfo
          nodes: Repository[]
        }
      }>(`
      query {
        search(first: ${PAGE_SIZE}, type: ${
        SearchType.Repository
      }, query: "org:${organizationName} ${queryString}"${getPageRequest()}) {
          repositoryCount
          pageInfo {
            endCursor
            hasNextPage
            hasPreviousPage
            startCursor
          }
          nodes {
            ... on Repository {
              name
              id
              pullRequests(first: 20, states: OPEN) {
                totalCount
                nodes {
                  number
                  title
                  commits (last: 1) {
                    totalCount
                    nodes {
                      commit {
                        checkSuites(last: 1) {
                          nodes {
                            conclusion
                          }
                        }
                      }
                    }
                  }
                  headRef {
                    name
                  }
                  author {
                    login
                  }
                }
              }
            }
          }
        }
      }
    `)
    },
    {
      keepPreviousData: true,
      onSuccess: (response) =>
        onSuccess(response.search.pageInfo, response.search.repositoryCount),
    },
  )

  if (isLoading) {
    return <div>loading...</div>
  }

  return (
    <div>
      <h1>Repositories of {organizationName}</h1>

      <label>
        Filter
        <input
          type={'text'}
          onChange={(event) => setQueryString(event.currentTarget.value)}
        />
      </label>

      <h4>
        Page {pages.currentPage + 1} of {pages.totalPages} (
        {data.search.repositoryCount} entries)
      </h4>

      {data.search.nodes.map((repository) => {
        return <RepositoryWithPrs key={repository.id} {...repository} />
      })}

      <button
        disabled={!pages[pages.currentPage]?.hasPreviousPage}
        onClick={prevPage}
      >
        Prev
      </button>
      <button
        disabled={!pages[pages.currentPage]?.hasNextPage}
        onClick={nextPage}
      >
        Next
      </button>
    </div>
  )
}
