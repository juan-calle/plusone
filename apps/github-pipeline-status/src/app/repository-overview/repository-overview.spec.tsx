import React from 'react'
import { render, screen, waitFor } from '@testing-library/react'
import { QueryClient, QueryClientProvider } from 'react-query'
import nock from 'nock'
import { BrowserRouter } from 'react-router-dom'

import { RepositoryOverview } from './repository-overview'

jest.mock('../octokit-provider/octokit-provider')
jest.mock('./repository-with-prs')

describe('RepositoryOverview', () => {
  const queryClient = new QueryClient({
    defaultOptions: { queries: { retry: 0, refetchOnWindowFocus: false } },
  })

  it('should render successfully', async () => {
    nock('https://api.github.com')
      .post('/graphql')
      .reply(200, {
        data: {
          search: {
            repositoryCount: 2,
            pageInfo: {
              endCursor: 'endCursor',
              hasNextPage: false,
              hasPreviousPage: false,
              startCursor: 'startCursor',
            },
            nodes: [
              {
                name: 'repo 1',
                id: '1',
              },
              { name: 'repo 2', id: '2' },
            ],
          },
        },
      })

    render(
      <QueryClientProvider client={queryClient}>
        <BrowserRouter>
          <RepositoryOverview />
        </BrowserRouter>
      </QueryClientProvider>,
    )

    await waitFor(() => {
      expect(nock.isDone()).toBeTruthy()
    })

    expect(screen.getByText(/repo 1/i)).toBeInTheDocument()
    expect(screen.getByText(/repo 2/i)).toBeInTheDocument()
  })
})
