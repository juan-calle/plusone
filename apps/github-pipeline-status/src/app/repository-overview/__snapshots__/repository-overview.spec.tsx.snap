// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryOverview should render successfully 1`] = `
<div>
  <div>
    <h1>
      Repositories of 
    </h1>
    <label>
      Filter
      <input
        type="text"
        value=""
      />
    </label>
    <table>
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Default Branch
          </th>
          <th>
            Open PRs
          </th>
          <th>
            ✓
          </th>
          <th>
            ✗
          </th>
          <th>
            other
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <a>
              repo 1
            </a>
          </td>
          <td>
            main
          </td>
          <td>
            <a
              href="undefined/pulls"
            >
              (
              1
               PRs)
            </a>
          </td>
          <td>
            1
          </td>
          <td />
          <td />
        </tr>
        <tr>
          <td>
            <a>
              repo 2
            </a>
          </td>
          <td>
            develop
          </td>
          <td />
          <td />
          <td />
          <td />
        </tr>
      </tbody>
    </table>
    <h4>
      Page 
      1
       of 
      1
       (
      2
       entries)
    </h4>
    <button
      disabled=""
    >
      Prev
    </button>
    <button
      disabled=""
    >
      Next
    </button>
  </div>
</div>
`;
